mapo/models/dynamics/gaussian.py

```diff
@@ -0,0 +1,80 @@
+# pylint: disable=missing-docstring
+
+import tensorflow as tf
+from tensorflow import keras
+from tensorflow.keras.layers import Dense
+import tensorflow_probability as tfp
+
+from ray.rllib.utils.annotations import override
+
+
+class GaussianDynamicsModel(keras.Model):
+    """GaussianDynamicsModel implements a state-action conditional
+    probabilistic transition model based on an unimodal Gaussian distribution
+    with parameterized mean and parameterized diagonal covariance matrix.
+
+    The Gaussian mean and diagonal covariance matrix are both parameterized by
+    feedforward networks with given 'layers' and 'activation'.
+
+    Args:
+        obs_space(gym.spaces.Box): The environment observation space.
+        action_space(gym.spaces.Box): The environment action space.
+
+    Keyword arguments:
+        layers(List[int]): A list of hidden layer units.
+        activation(tf.Tensor): An activation function op.
+    """
+
+    def __init__(self, obs_space, action_space, **kwargs):
+        super(GaussianDynamicsModel, self).__init__()
+
+        self.obs_space = obs_space
+        self.action_space = action_space
+
+        self.config = {**kwargs}
+
+        self.hidden_layers = []
+        activation = self.config.get("activation", "relu")
+        for units in self.config["layers"]:
+            self.hidden_layers.append(Dense(units, activation=activation))
+
+        self.mean_output_layer = Dense(self.obs_space.shape[0])
+        self.log_stddev_output_layer = Dense(self.obs_space.shape[0])
+
+    @override(keras.models.Model)
+    def call(self, inputs, training=None, mask=None):
+        """Returns the 2-head feedforward network output implementing
+        the Gaussian distribution's mean and log std deviation.
+
+        Args:
+            inputs(Tuple(tf.Tensor, tf.Tensor)): A pair of (state, action) tensors.
+
+        Return:
+            Tuple(tf.Tensor, tf.Tensor): A pair of tensors (mean, log_stddev).
+        """
+        inputs = keras.layers.concatenate(inputs)
+        for layer in self.hidden_layers:
+            inputs = layer(inputs)
+        outputs = (self.mean_output_layer(inputs), self.log_stddev_output_layer(inputs))
+        return outputs
+
+    def dist(self, state, action):
+        """Returns the next state distribution conditioned on the state
+        and the action."""
+        mean, log_stddev = self.call([state, action])
+        gaussian = tfp.distributions.Normal(loc=mean, scale=tf.exp(log_stddev))
```

Comment:
Why not use [MultivariateNormalDiag](https://www.tensorflow.org/probability/api_docs/python/tfp/distributions/MultivariateNormalDiag) here?

Reply:
I guess we could but I decided to go for a simpler model to start for the initial experiments... but I will keep this in mind.